fs/io_zenfs.cc

```diff
@@ -632,6 +633,171 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
   return zoneFile_->GetUniqueId(id, max_size);
 }
 
+// This is a helper function to read data from a source zone from a read
+// position -> read_pos.
+IOStatus ZenFSGCWorker::ReadExtent(Slice* buf, uint64_t read_pos,
+                                   Zone* zone_src) {
+  int f = zbd_->GetReadFD();
+  const char* data = buf->data();
+  size_t read = 0;
+  size_t to_read = buf->size();
+  int ret;
+
+  if (read_pos >= zone_src->wp_) {
+    // EOF
+    buf->clear();
+    return IOStatus::OK();
+  }
+
+  if ((read_pos + to_read) > (zone_src->start_ + zone_src->max_capacity_)) {
+    return IOStatus::IOError("Read across zone");
+  }
+
+  while (read < to_read) {
+    ret = pread(f, (void*)(data + read), to_read - read, read_pos);
+
+    if (ret == -1 && errno == EINTR) continue;
+    if (ret < 0) return IOStatus::IOError("Read failed");
+
+    read += ret;
+    read_pos += ret;
+  }
+
+  return IOStatus::OK();
+}
+
+// This is a heavy weight function. There is going to be a high
+// traffic activity via the PCIe channel to the ZNS SSD because
+// of the read/write to zones which needs  to be issued. We need
+// some better ideas later to bring in efficiency, something like
+// "simple copy" or ideas in those lines.
+IOStatus ZenFSGCWorker::MoveValidDataToNewDestZone() {
+  std::vector<Zone*>::iterator zone_it;
+  std::vector<ZoneExtent*>::iterator ext_it;
+  std::vector<ZoneExtent*>::iterator e_it;
+  IOStatus s;
+  uint64_t r_pos;
+  uint32_t size;
+  uint32_t long_ext_size;
+  uint64_t new_start;
+  const char* ptr;
+  int dont_read = 0;
+
+  // Sort the Extent list in decreasing order.
+  std::sort(extent_list.begin(), extent_list.end(),
+            [](ZoneExtent* ext1, ZoneExtent* ext2) {
+              return ext1->length_ > ext2->length_;
+            });
+
+  // Get the size of the largest extent.
+  long_ext_size = extent_list[0]->length_;
+
+  // Allocate a buffer with size of the largest extent.
+  // We have to issue pread from the source zones so we
+  // need a buffer.
+  ptr = new char[long_ext_size];
+
+  zone_it = dst_zone_list.begin();
+  for (ext_it = extent_list.begin(); ext_it != extent_list.end();) {
+    ZoneExtent* ext;
+    Zone* zone_dst;
+
+    ext = *ext_it;
+    zone_dst = *zone_it;
+
+    // Set the position and length in the source zone to
+    // read the data.
+    r_pos = ext->start_;
+    size = ext->length_;
+    Slice buf(ptr, size);
+
+    if (!dont_read) s = ReadExtent(&buf, r_pos, ext->zone_);
+
+    // Free the allocated memory if error.
+    if (!s.ok()) {
+      if (ptr) delete[] ptr;
+      return s;
+    }
+
+    // Store the new starting position for the extent
+    // which will be later made persistent.
+    new_start = zone_dst->wp_;
```

Comment:
If there could be multiple threads appending data to a zone, how could we ensure that `new_start` is really at `zone_dst->wp_`?

Reply:
> If there could be multiple threads appending data to a zone, how could we ensure that `new_start` is really at `zone_dst->wp_`?

Yes that's a valid concern. We need to figure out a way to serialize that. I did not find a way to lock the zone and operate on it or maybe there is a way. Perhaps we need a bit more thought on this.

Reply:
Yes. As far as I can see, in ZenFS, there is only one writable file in one zone. `AllocateZone` will mark a zone as `being written` after allocating, making this zone not being able to used by other files for write. Maybe allocating free zone in `dst_zone_list` could solve this issue.

Reply:
> Yes. As far as I can see, in ZenFS, there is only one writable file in one zone. `AllocateZone` will mark a zone as `being written` after allocating, making this zone not being able to used by other files for write. Maybe allocating free zone in `dst_zone_list` could solve this issue.

Hmm. OK, we have to add that logic in the function GetDestZoneToMoveValidData()  and the function should push zones to the dst_zone_list member variable.

Reply:
> > Yes. As far as I can see, in ZenFS, there is only one writable file in one zone. `AllocateZone` will mark a zone as `being written` after allocating, making this zone not being able to used by other files for write. Maybe allocating free zone in `dst_zone_list` could solve this issue.
> 
> Hmm. OK, we have to add that logic in the function GetDestZoneToMoveValidData() and the function should push zones to the dst_zone_list member variable.
- For getting the list of destination zones, in the function GetDestZoneToMoveValidData(), we may need to call AllocateZone() with open_for_write_ set to false or write another version of AllocateZone() ex. AllocateZoneForGC() which sets that flag to false. Then in the function MoveValidDataToNewDestinationZone(), after we fetch the current parameters of the destination zones like wp_ (write pointer), we can set the flag  open_for_write_ back to true and then start appending to that zone.

Reply:
As in https://github.com/bzbd/zenfs/blob/master/fs/zbd_zenfs.cc#L518, allocate a zone without setting `open_for_write` to true may lead to some issues. The zone might be seen in "open but not allocated" state, and will be allocated to other files. Therefore, I don't think this would work.